Updated invite route and function for new standards, and added tests.

TheHollidayInn opened the pull request:
Here I have updated the private functions to use new syntax. I have also implemented the ability to invite by email and userId. Then, I added tests. 

This PR is what I need merged before adding tests and the other group routes.

crookedneighbor commented on test/api/v3/integration/groups/POST-groups_invite.test.js:
```diff
@@ -0,0 +1,198 @@
+
+  beforeEach(async () => {
+    user = await generateUser({balance: 1});
+    let groupName = 'Test Public Guild';
```
I don't think it's meaningful to pull these out into variables.

crookedneighbor commented on test/api/v3/integration/groups/POST-groups_invite.test.js:
```diff
@@ -0,0 +1,198 @@
+    });
+
+    it('invites a user to a group by uuid', async () => {
+      let newUser = await generateUser();
```
Let's be more meaningful with the variable name. Maybe something like `invitedUser`

crookedneighbor commented on test/api/v3/integration/groups/POST-groups_invite.test.js:
```diff
@@ -0,0 +1,198 @@
+      let invite = await user.post(`/groups/${group._id}/invite`, {
+        uuids: [newUser._id],
+      });
+      let invitedUser = await newUser.get('/user');
```
This could instead be:

```
await expect(invitedUser.get('/user')).to.eventually.have.deep.property('invitedUser.invitations.guilds[0].id', group._id);
```

crookedneighbor commented on test/api/v3/integration/groups/POST-groups_invite.test.js:
```diff
@@ -0,0 +1,198 @@
+} from '../../../../helpers/api-integration.helper';
+
+describe('Post /groups/:groupId/invite', () => {
+  let user;
```
Let's make this something like `inviter` or `invitingUser`

crookedneighbor commented on test/api/v3/integration/groups/POST-groups_invite.test.js:
```diff
@@ -0,0 +1,198 @@
+
+  describe('user id invites', () => {
+    it('returns an error when invited user is not found', async () => {
+      let fakeID = 'fakeID';
```
This should probably be a uuid instead of the string `'fakeID'`.

crookedneighbor commented on test/api/v3/integration/groups/POST-groups_invite.test.js:
```diff
@@ -0,0 +1,198 @@
+
+      expect(invite).to.exist;
+      expect(invitedUser.invitations.guilds[0].id).to.equal(group._id);
+      expect(invitedUser2.invitations.guilds[0].id).to.equal(group._id);
```
Same two notes as the last test, could have more meaningful variable names and you can write the expectation directly on the `get('/user')` call for each user.

crookedneighbor commented on test/api/v3/integration/groups/POST-groups_invite.test.js:
```diff
@@ -0,0 +1,198 @@
+      });
+      let invitedUser = await newUser.get('/user');
+
+      expect(invite).to.exist;
```
There should be a separate test that asserts on what is returned from the invite call.

crookedneighbor commented on test/api/v3/integration/groups/POST-groups_invite.test.js:
```diff
@@ -0,0 +1,198 @@
+        emails: [{name: 'test', email: 'test@habitca.com'}],
+      });
+
+      expect(invite).to.exist;
```
We should decide explicitly what the invite call returns and assert on it.

crookedneighbor commented on test/api/v3/integration/groups/POST-groups_invite.test.js:
```diff
@@ -0,0 +1,198 @@
+        error: 'NotAuthorized',
+        message: t('userAlreadyInvitedToGroup'),
+      });
+    });
```
What happens when you invite multiple users and one is already in the group/invited to the group and the others are not?

TheHollidayInn replied:
It should just throw an error as shown in a test above. I'll make an explicit test for multiple users. Throwing an error in the middle may not be the best though, because, some users will have invites. Hmm..

crookedneighbor commented on test/api/v3/integration/groups/POST-groups_invite.test.js:
```diff
@@ -0,0 +1,198 @@
+
+  describe('party invites', () => {
+    let party;
+    let partyName = 'Test Party';
```
Again, not meaningful to have these as variable names.

crookedneighbor commented on website/src/controllers/api-v3/groups.js:
```diff
@@ -391,145 +392,101 @@ api.removeGroupMember = {
+    if (userToInvite.party._id) {
+      throw new NotAuthorized(res.t('userAlreadyInAParty'));
+    }
+    // @TODO: Why was this here?
```
Can you explain your question better?

crookedneighbor commented on website/src/controllers/api-v3/groups.js:
```diff
@@ -563,15 +520,25 @@ api.inviteToGroup = {
     let uuids = req.body.uuids;
     let emails = req.body.emails;
```
Should we have a cap on the number of uuids and emails a user can invite at once?

paglias replied:
Hmm probably a good idea. At some point we'll also have to enforce limits in other places too

TheHollidayInn replied:
That's a good idea. What should I cap at?

paglias replied:
I'd say 100

crookedneighbor commented on website/src/controllers/api-v3/groups.js:
```diff
@@ -563,15 +520,25 @@ api.inviteToGroup = {
-    } else if (Array.isArray(emails)) {
-      // return _inviteByEmails(emails, group, user, req, res, next);
-    } else {
+    if (!Array.isArray(uuids) && !Array.isArray(emails)) {
```
Could we do something like:

``` js
let hasUuids = !_.isEmpty(Array.isArray(uuids));
let hasEmails = !_.isEmpty(Array.isArray(emails));

if (!hasUuids && !hasEmails) {
  throw new BadRequest(res.t('canOnlyInviteEmailUuid'));
}

let results = [];

if (hasUuids) {
  // stuff
}

if (hasEmails) {
  // stuff
}
```

paglias replied:
why `_.isEmpty` and not just `hasUuids = Array.isArray(uuids)` and `if (uuids)`?

TheHollidayInn replied:
How about this?

```
    let uuids = req.body.uuids;
    let emails = req.body.emails;

    if (!uuids && !emails) {
      throw new BadRequest(res.t('canOnlyInviteEmailUuid'));
    }

    let uuidsIsArray = Array.isArray(uuids);
    let emailsIsArray = Array.isArray(emails);

    let results = [];

    if (uuidsIsArray) {
      let uuidInvites = uuids.map((uuid) => _inviteByUUID(uuid, group, user, req, res));
      let uuidResults = await Q.all(uuidInvites);
      results.push(uuidResults);
    }

    if (emailsIsArray) {
      let emailInvites = emails.map((invite) => _inviteByEmail(invite, group, user, req, res));
      let emailResults = await Q.all(emailInvites);
      results.push(emailResults);
    }
```

paglias replied:
@TheHollidayInn seems good but since we only support emails OR uuids there's no need to push emailResults to results, just do `results = await Q.all(emailInvited/uuidInvites)`

TheHollidayInn replied:
Well, I updated the code to support emails and uuids.

paglias replied:
@TheHollidayInn at the same time? In which case this check should be removed https://github.com/HabitRPG/habitrpg/pull/6475/files#diff-110878cdf614e13dc8eac3ea77ed6265R523

TheHollidayInn replied:
Alright, I'm going to add two more tests for each email and uuids to test if there was an object passed and if the array is empty. Then, I'll test out Blades code to see if it passes. Does that sounds correct? I'll post an update soon if you need to see.

TheHollidayInn replied:
And, to clarify, here is the logic I am about to refactor the code for:

//If we have neither, return error
//If we have uuids, but empty or object, return error
 //If we have emails, but empty or object, return error

paglias replied:
Yeah it looks good, I misread the code. One thing that may change is the error string here https://github.com/HabitRPG/habitrpg/pull/6475/files#diff-110878cdf614e13dc8eac3ea77ed6265R524 

And when you push to results you may want to use the spread operator to push invites individually to `results`: `results.push(...uuidResults)` More infot https://developer.mozilla.org/it/docs/Web/JavaScript/Reference/Operators/Spread_operator

TheHollidayInn replied:
Yea, I just saw that error again while refactoring. I'm going to change it for sure.

And, sweet. I'll look into that operator.

crookedneighbor commented on website/src/controllers/api-v3/groups.js:
```diff
@@ -563,15 +520,25 @@ api.inviteToGroup = {
+      results.push(emailResults);
+    }
+
+    res.respond(200, {});
```
We should respond with something. Maybe: `{ message: "<number> invitations sent" }`

paglias replied:
or return the invitations that are now on the users' profiles

paglias replied:
@TheHollidayInn regarding this one it would be very useful to return the invitations that are stored on the users so we can update the UI without a new request

TheHollidayInn replied:
Will do!

crookedneighbor commented on website/src/controllers/api-v3/groups.js:
```diff
@@ -391,145 +392,101 @@ api.removeGroupMember = {
+  } else {
+    userReturnInfo = invite.email;
+    // yeah, it supports guild too but for backward compatibility we'll use partyInvite as query
+    // TODO absolutely refactor this horrible code
```
We should listen to this TODO :smile:

TheHollidayInn replied:
Yea, I move some of the strings out to make it more readable. I'll take a look to see if there is any more to change.

crookedneighbor commented on website/src/controllers/api-v3/groups.js:
```diff
@@ -391,145 +392,101 @@ api.removeGroupMember = {
+
+    let variables = [
+      {name: 'LINK', content: link},
+      {name: 'INVITER', content: inviter || inviter.name},
```
This doesn't look quite right to me. User is getting passed in here, so it should be `inviter.profile.name`

crookedneighbor commented on website/src/controllers/api-v3/groups.js:
```diff
@@ -391,145 +392,101 @@ api.removeGroupMember = {
+    let variables = [
+      {name: 'LINK', content: link},
+      {name: 'INVITER', content: inviter || inviter.name},
+      {name: 'REPLY_TO_ADDRESS', content: inviter.email},
```
This is also not right.

crookedneighbor commented on website/src/controllers/api-v3/groups.js:
```diff
@@ -391,145 +392,101 @@ api.removeGroupMember = {
+    // TODO absolutely refactor this horrible code
+    const partyQueryString = JSON.stringify({id: group._id, inviter, name: group.name});
+    const encryptedPartyqueryString = encrypt(partyQueryString);
+    let link = `?partyInvite=${encryptedPartyqueryString}`;
```
@paglias Is this `partyInvite` query something in our email server?

paglias replied:
No it's not on the email server, it's used on the main server to show invitations to non-registered users. If you open a habitica.com?partyInvite=code link I think once you signup you automatically get invited to the group. This is the old code https://github.com/HabitRPG/habitrpg/blob/develop/website/src/controllers/api-v2/user.js#L547 I'll refactor this once we get to that point

TheHollidayInn committed:
Updated invite route and function for new standards, and added tests.

TheHollidayInn commented:
Okay, updates:
-I removed extra abstract variables
-I added more meaningful variables names
-Added real UUIDs for examples
-Added limit and tests for invites - to be change per our slack discussion
-And, I added a test that when a user invites multiple users (and on user has an error) that the server will still respond with an error. I'm unsure if this is the functionality we want though.

paglias commented on website/src/controllers/api-v3/groups.js:
```diff
@@ -391,145 +395,105 @@ api.removeGroupMember = {
-        });
-      }
+async function _inviteByUUID (uuid, group, inviter, req, res) {
+  // @TODO: Add Push Notifications
```
@TheHollidayInn there's a reason why the push notifications code was removed?

paglias commented on website/src/controllers/api-v3/groups.js:
```diff
@@ -391,145 +395,105 @@ api.removeGroupMember = {
+    if (_.find(userToInvite.invitations.guilds, {id: group._id})) {
+      throw new NotAuthorized(res.t('userAlreadyInvitedToGroup'));
+    }
+    userToInvite.invitations.guilds.push({id: group._id, name: group.name, inviter: res.locals.user._id});
```
@TheHollidayInn instead of `inviter: res.locals.user._id` you could use `inviter: inviter._id`

TheHollidayInn committed:
Added invite limit and tests for when multiple users with an error
